File: Frontend/src/services/study-plan.ts
import { apiClient } from "./api-client";

/** Matches Backend StudyPlanResponse. */
export interface StudyPlan {
  id: number;
  student_id: number;
  goal_id: number;
  subject_id?: number | null;
  title: string;
  task_description?: string | null;
  rag_content?: string | null;
  lesson_summary?: string | null;
  generation_status: "not_started" | "queued" | "generating" | "ready" | "failed";
  lesson_status: "not_started" | "queued" | "generating" | "ready" | "failed";
  quiz_status: "not_started" | "queued" | "generating" | "ready" | "failed";
  generation_error?: string | null;
  generation_attempts: number;
  generation_started_at?: string | null;
  generation_finished_at?: string | null;
  study_date: string;
  start_time: string;
  end_time: string;
  ai_generated?: boolean;
  status: "todo" | "doing" | "done";
  created_at: string;
}

export const studyPlanApi = {
  getPlans: (params?: {
    goal_id?: number;
    status_filter?: string;
    start_date?: string;
    end_date?: string;
  }) => apiClient.get<StudyPlan[]>("/plans/", { params }),

  getPlan: (planId: number) => apiClient.get<StudyPlan>(`/plans/${planId}`),

  startGeneration: (planId: number) =>
    apiClient.post<StudyPlan>(`/plans/${planId}/generation`),

  retryGeneration: (planId: number) =>
    apiClient.post<StudyPlan>(`/plans/${planId}/generation/retry`),

  updatePlanStatus: (planId: number, status: "todo" | "doing" | "done") =>
    apiClient.patch<StudyPlan>(`/plans/${planId}`, { status }),
};

export default studyPlanApi;

File: Backend/app/schemas/study_plan.py
from datetime import date, datetime, time
from typing import Optional

from pydantic import BaseModel


# ── Tạo mới study plan (dùng bởi plan_repository & services) ─
class StudyPlanCreate(BaseModel):
    student_id: int
    goal_id: int
    title: str
    task_description: Optional[str] = None
    rag_content: Optional[str] = None
    lesson_summary: Optional[str] = None
    generation_status: str = "not_started"
    lesson_status: str = "not_started"
    quiz_status: str = "not_started"
    generation_error: Optional[str] = None
    generation_attempts: int = 0
    generation_started_at: Optional[datetime] = None
    generation_finished_at: Optional[datetime] = None
    study_date: date
    start_time: time
    end_time: time
    ai_generated: bool = False
    status: str = "todo"


# ── Response từng study plan trả về cho client ───────────────
class StudyPlanResponse(BaseModel):
    id: int
    student_id: int
    goal_id: int
    subject_id: Optional[int] = None
    title: str
    task_description: Optional[str] = None
    rag_content: Optional[str] = None
    lesson_summary: Optional[str] = None
    generation_status: str
    lesson_status: str
    quiz_status: str
    generation_error: Optional[str] = None
    generation_attempts: int
    generation_started_at: Optional[datetime] = None
    generation_finished_at: Optional[datetime] = None
    study_date: date
    start_time: time
    end_time: time
    ai_generated: bool
    status: str  # "todo" | "doing" | "done"
    created_at: datetime

    class Config:
        from_attributes = True


# ── Cập nhật trạng thái task ─────────────────────────────────
class StudyPlanUpdate(BaseModel):
    status: Optional[str] = None  # "todo" | "doing" | "done"
    title: Optional[str] = None
    task_description: Optional[str] = None
    rag_content: Optional[str] = None
    study_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
